test/tools/runner/hooks/leak_checker.ts

```diff
@@ -0,0 +1,179 @@
+/* eslint-disable @typescript-eslint/no-this-alias */
+import { expect } from 'chai';
+import * as chalk from 'chalk';
+import { Socket } from 'net';
+
+import { MongoClient } from '../../../../src/mongo_client';
+import { ServerSessionPool } from '../../../../src/sessions';
+
+class LeakChecker {
+  static originalAcquire: typeof ServerSessionPool.prototype.acquire;
+  static originalRelease: typeof ServerSessionPool.prototype.release;
+  static kAcquiredCount: symbol;
+
+  static originalConnect: typeof MongoClient.prototype.connect;
+  static originalClose: typeof MongoClient.prototype.close;
+  static kConnectCount: symbol;
+
+  static {
+    LeakChecker.originalAcquire = ServerSessionPool.prototype.acquire;
+    LeakChecker.originalRelease = ServerSessionPool.prototype.release;
+    LeakChecker.kAcquiredCount = Symbol('acquiredCount');
+
+    LeakChecker.originalConnect = MongoClient.prototype.connect;
+    LeakChecker.originalClose = MongoClient.prototype.close;
+    LeakChecker.kConnectCount = Symbol('connectedCount');
+  }
+
+  clients: Set<MongoClient>;
+  sessionPools: Set<ServerSessionPool>;
+
+  constructor(public titlePath: string) {
+    this.clients = new Set<MongoClient>();
+    this.sessionPools = new Set<ServerSessionPool>();
+  }
+
+  setupSessionLeakChecker() {
+    const leakChecker = this;
+    ServerSessionPool.prototype.acquire = function (...args) {
+      leakChecker.sessionPools.add(this);
+
+      if (!(LeakChecker.kAcquiredCount in this)) {
+        this[LeakChecker.kAcquiredCount] = 1;
+      } else {
+        this[LeakChecker.kAcquiredCount] += 1;
+      }
```

Comment:
superficial nit suggestion:

```suggestion
      this[LeakChecker.kAcquiredCount] = this[LeakChecker.kAcquiredCount] ?? 0
      this[LeakChecker.kAcquiredCount] += 1
```

Reply:
If you decide to make this change - I think it makes sense to apply everywhere we're creating a default value for a new property on the object

Reply:
done